### DISCUSSION TITLE
WSL auto-attach

### DISCUSSION THREAD

**Discussion by DrInfiniteExplorer** [dorssel/usbipd-win#168] [General]
Hi!
To start with, big thanks for developing this!

I've recently started fiddling around with usbip-win & WSL and have gotten some stuff working after doing
a) removing that filter (registered by wireshark)
b) getting wsl-side `usbip` going
c) pulling new kernel from Microsoft
d) building the kernel, with drivers!
e) figuring out that I was missing udev-rules
f) inadvertently discovering that a upgrade to Ubuntu20.04 added udev-rules
g) realizing that the udev-service needs to be manually kickstarted
h) adapting the solution [here](https://github.com/microsoft/WSL/issues/4166#issuecomment-618159162) (point 2&3) to do `sudo service udev restart` if needed

Now I'm pretty happy with the situation!
Except that I need to reattach the device if it restarts, disconnects, or is physically disconnected.
Which happens quite a bit when I'm programming devices or communicating with my 3D-printer running klipper.

So I thought I'd mangle the example [here](https://docs.microsoft.com/en-us/windows/win32/devio/registering-for-device-notification) a bit to make it auto-attach devices of my choosing as they are registered in windows.
But before I even got started I realized that this would be a fancy feature to add to _usbipd-win_ itself!

Do you think this would be a viable idea?
Would you be open to a PR that adds the functionality?
It wouldn't automagically fix the case of (re)attaching devices on WSL-startup, but a WSL-instance that is starting could run `usbipd.exe wsl --wsl-is-starting-reattach-autoattach-devices-please` on the windows side ¯\_(ツ)_/¯

Would it be alright to add a note about some of those issues I mentioned to the wiki or docs?
(How does one send PRs for the wiki-git? Seems to be unsupported by github)

**Comment by dorssel** [comment]
Thanks for the elaborate report.

a) Will be solved by #164 
b,c,d) already covered by WSL wiki
e) duplicate of #121 
f) new info, thanks
g,h) duplicate of #60 (already marked with Documentation)

WSL docs are mostly handled by Microsoft (@benmcmorran)

**Reply by DrInfiniteExplorer** [reply]
I kind of get the feeling that you didn't read the whole post.. 😅

I've now put together an abomination consisting of a bit of c++ that invokes `py attach.py {VID:PID}` using `system` on detection of usb-serials, and `attach.py` invokes a nested command in wsl-bash to figure out the busid given a {VID:PID} and attaching to that. Works like a charm!
Would obviously be better to do it directly from usbipd-win directly 😛 

#164 as far as I understand that, it would force the serial to be bound to either (general purpose windows) or exposure through uspipd-win - I personally wouldn't necessarily want to go that route if I don't have to; Adding some docs about the filter (with workaround like unregistering USBPcap, _or_ workaround using ´--force`) would be awesome.

Should I add my h) -solution to #60 for posterity? Given the title of #60 I probably never would have realized that it was related to udev/access -troubles 😅

**Reply by dorssel** [reply]
O, I did read all of it. You touched so many subjects in one post that I just forgot to mention that that bit is actually a duplicate of #56.

Other workarounds I have seen include a simple attach-loop from within WSL with a second of pause in-between.

The auto attach idea is nice. It's on the backlog.

**Comment by Ashram56** [comment]
Trying to unravel a similar situation, and I'm not sure I found a solution for this. This has probably been asked on multiple occasions, apologies if this is documented somewhere else.

Essentially I have a flashing tool for an embedded device connected through USB that runs inside a docker container. During the process, it will reboot the target device once, which causes a USBip disconnect and make the process fail

I used to use VMWare Workstation which can trigger a reconnect at reboot of USB device, but I can't find a way to do so with WSL2 and USBIP. I thought the "--force" flag would work, but I did not really find a documentaiton for it. If I use "bind" it reports an error stating "device has already been attached" (even though it's actually not attached when I run it)

Thanks for any insight you can share

Regards

**Reply by DrInfiniteExplorer** [reply]
@Ashram56 
There is a twofold problem: Docker dropping mounted devices, and auto-reattachment.

From what I gathered last time I fiddled with it, dropping the mounted device(without ability to auto-remount without restart) is a flaw in all Docker systems.
People on the internet said _"Just run the docker with `--privileged` lol `¯\_(ツ)_/¯`"_ which I found to not be very helpful. And some people reported that even when doing that, the re-mounted device will only be visible in OTHER docker instances (which are not even running as privileged, smh my head)
I don't know if it's possible on other systems, but I tried re-mounting / running USBIP inside the docker (inside WSL) itself, but that didn't help any, probably because in the end it's the same kernel that the USBIP-device is mounted inside. Mounting it from inside docker only made it appear in (outer WSL space) and not in the docker itself. (Which is kinda a flaw with Docker since one of the cool sellers is to not escape the jail? Or a problem of the kernel? Or something? IDK).

[Here](https://github.com/DrInfiniteExplorer/shitty_usbipd-win_wsl_autoattacher) is my crappy auto-attacher, which at least can automate the WSL-side attaching. I found that for _some_ simple FW stuff that resets occasionally, it manages to reconnect things without Docker dropping the mount about 60% of the times.
My crappy auto-attacher would probably need adaption to recognize the device you want to attach, but you should be able to run it and display whatever pops up in the output to identify your device and the make the appropriate adjustments ¯\_(ツ)_/¯

In the end I prefer to not run with `--privileged` and the auto-attacher works OK often enough for me to not waste more energy on that.

Dunno if any of this is what you asked about, hope you have a neat day regardless

**Comment by dorssel** [comment]
See https://github.com/dorssel/usbipd-win/issues/56#issuecomment-1101830389 for auto-attach.
